lib/libzfs_core/libzfs_core.c

```diff
@@ -645,6 +646,58 @@ lzc_send_resume(const char *snapname, const char *from, int fd,
 	    resumeoff, NULL));
 }
 
+struct sendargs {
+	int ioctlfd;
+	int inputfd;
+	int outputfd;
+};
+typedef struct sendargs sendargs_t;
+
+void *
+__do_send_output(void *voidargs)
+{
+	sendargs_t *args = (sendargs_t *)voidargs;
+	sigset_t sigs;
+	int buflen = 131072;
+
+	// see the comment above the close() calls for why
+	// we can't just die from SIGPIPE
+	sigemptyset(&sigs);
+	sigaddset(&sigs, SIGPIPE);
+	pthread_sigmask(SIG_BLOCK, &sigs, NULL);
+
+#ifndef __linux__
+	void* buf = calloc(1, buflen);
+#endif
+
+	int err = 0;
+	while (err >= 0) {
+#ifdef __linux__
+		err = splice(args->inputfd, 0, args->outputfd, 0, buflen,
+		    SPLICE_F_MORE);
+#else
+		err = read(args->inputfd, buf, buflen);
+		if (err <= 0) {
+			break;
+		}
+		err = write(args->outputfd, buf, err);
+#endif
+		if (err <= 0) {
+			break;
```

Comment:
This could be folded into the while condition if you make it `err > 0` and initialise to 1

Reply:
Makes sense. Done.